src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -426,15 +436,22 @@ private void WriteFields(TypeBuilderImpl typeBuilder)
                     AddFieldLayout(handle, field._offset);
                 }
 
-                if (field._marshallingData != null)
+                if (field.Attributes.HasFlag(FieldAttributes.HasFieldMarshal) && field._marshallingData != null)
                 {
                     AddMarshalling(handle, field._marshallingData.SerializeMarshallingData());
                 }
 
-                if (field._defaultValue != DBNull.Value)
+                if (field.Attributes.HasFlag(FieldAttributes.HasDefault) && field._defaultValue != DBNull.Value)
                 {
                     AddDefaultValue(handle, field._defaultValue);
                 }
+
+                if (field.Attributes.HasFlag(FieldAttributes.HasFieldRVA) && field._rvaSize > 0)
+                {
+                    _metadataBuilder.AddFieldRelativeVirtualAddress(handle, fieldDataBuilder.Count);
+                    fieldDataBuilder.WriteBytes(field._rvaData!);
+                    fieldDataBuilder.Align(8);
```

Comment:
Is there a constant for the 8?

Reply:
The constant defined within internal type `ManagedTextSection.cs`
https://github.com/dotnet/runtime/blob/1be0fb324d37c4672ef9cef86254ce0214100f0b/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/ManagedTextSection.cs#L122

Reply:
Ah, just found a public API in ManagedPRBuilder https://learn.microsoft.com/en-us/dotnet/api/system.reflection.portableexecutable.managedpebuilder.mappedfielddataalignment?view=net-7.0